source/legacy_ingress.go

```diff
@@ -0,0 +1,304 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"text/template"
+
+	log "github.com/sirupsen/logrus"
+	extensionsv1beta1 "k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/labels"
+	kubeinformers "k8s.io/client-go/informers"
+	extinformers "k8s.io/client-go/informers/extensions/v1beta1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+// legacyIngressSource is an implementation of Source for Kubernetes ingress objects.
+// Ingress implementation will use the spec.rules.host value for the hostname
+// Use targetAnnotationKey to explicitly set Endpoint. (useful if the ingress
+// controller does not update, or to override with alternative endpoint)
+type legacyIngressSource struct {
+	client                   kubernetes.Interface
+	namespace                string
+	annotationFilter         string
+	fqdnTemplate             *template.Template
+	combineFQDNAnnotation    bool
+	ignoreHostnameAnnotation bool
+	ingressInformer          extinformers.IngressInformer
+	ignoreIngressTLSSpec     bool
+	ignoreIngressRulesSpec   bool
+	labelSelector            labels.Selector
+}
+
+// NewLegacyIngressSource creates a new ingressLegacySource with the given config.
+func NewLegacyIngressSource(ctx context.Context, kubeClient kubernetes.Interface, namespace, annotationFilter string, fqdnTemplate string, combineFqdnAnnotation bool, ignoreHostnameAnnotation bool, ignoreIngressTLSSpec bool, ignoreIngressRulesSpec bool, labelSelector labels.Selector) (Source, error) {
```

Comment:
should the return sig be `(legacyIngressSource, error)`?

Reply:
`Source` is an interface that is implemented by `legacyIngressSource`